### PR TITLE
[SHELL32] SetNameOf should not append old extension to folders

### PR DESCRIPTION
**Pull request by whindsaks**

### CODE REVIEW

File: dll/win32/shell32/folders/CFSFolder.cpp [hunk 1]
```diff
@@ -1499,47 +1499,42 @@ HRESULT WINAPI CFSFolder::SetNameOf(
...
+    PathCombineW(szSrc, m_sPathTarget, pDataW->wszName); // FIXME: PIDLs without wide string
 
     /* build destination path */
     if (dwFlags == SHGDN_NORMAL || dwFlags & SHGDN_INFOLDER)
         PathCombineW(szDest, m_sPathTarget, lpName);
     else
         lstrcpynW(szDest, lpName, MAX_PATH);
 
-    if(!(dwFlags & SHGDN_FORPARSING) && SHELL_FS_HideExtension(szSrc)) {
+    if(!(dwFlags & SHGDN_FORPARSING) && !bIsFolder && SHELL_FS_HideExtension(szSrc)) {
```
Comment (katahiromz) [inline dll/win32/shell32/folders/CFSFolder.cpp:1510]:
```suggestion
    if (!(dwFlags & SHGDN_FORPARSING) && !bIsFolder && SHELL_FS_HideExtension(szSrc)) {
```

File: dll/win32/shell32/folders/CFSFolder.cpp [hunk 2]
```diff
@@ -1499,47 +1499,42 @@ HRESULT WINAPI CFSFolder::SetNameOf(
...
         WCHAR *ext = PathFindExtensionW(szSrc);
         if(*ext != '\0') {
             INT len = wcslen(szDest);
             lstrcpynW(szDest + len, ext, MAX_PATH - len);
         }
```
Comment (katahiromz) [inline dll/win32/shell32/folders/CFSFolder.cpp:1515]:
```suggestion
        if (*ext)
            PathAddExtensionW(szDest, ext);
```

File: dll/win32/shell32/folders/CFSFolder.cpp [hunk 3]
```diff
@@ -1499,47 +1499,42 @@ HRESULT WINAPI CFSFolder::SetNameOf(
...
 
     /* build destination path */
     if (dwFlags == SHGDN_NORMAL || dwFlags & SHGDN_INFOLDER)
         PathCombineW(szDest, m_sPathTarget, lpName);
     else
         lstrcpynW(szDest, lpName, MAX_PATH);
 
-    if(!(dwFlags & SHGDN_FORPARSING) && SHELL_FS_HideExtension(szSrc)) {
+    if(!(dwFlags & SHGDN_FORPARSING) && !bIsFolder && SHELL_FS_HideExtension(szSrc)) {
         WCHAR *ext = PathFindExtensionW(szSrc);
```
Comment (katahiromz) [inline dll/win32/shell32/folders/CFSFolder.cpp:1511]:
```suggestion
        LPWSTR ext = PathFindExtensionW(szSrc);
```

**Review by binarymaster — COMMENTED** [review]
Is there API tests for that function? (if not, I suggest covering that)

**Comment by whindsaks** [comment]
> Is there API tests for that function? (if not, I suggest covering that)

The relevant setting is not on by default. `SHGetSetSettings` has not even been implemented yet so we can turn it on and off correctly.

File: dll/win32/shell32/folders/CFSFolder.cpp [hunk 4]
```diff
@@ -1499,47 +1499,41 @@ HRESULT WINAPI CFSFolder::SetNameOf(
     }
 
     /* build source path */
-    PathCombineW(szSrc, m_sPathTarget, pDataW->wszName);
+    PathCombineW(szSrc, m_sPathTarget, pDataW->wszName); // FIXME: PIDLs without wide string
```
Comment (HBelusca) [inline dll/win32/shell32/folders/CFSFolder.cpp:1502]: What do you mean? Do you mean that what's stored in wszName is not a wide string? or?

File: dll/win32/shell32/folders/CRegFolder.cpp [hunk 5]
```diff
@@ -772,8 +772,10 @@ HRESULT WINAPI CRegFolder::SetNameOf(HWND hwndOwner, PCUITEMID_CHILD pidl,    /*
 
     if (res == ERROR_SUCCESS)
     {
-        *pPidlOut = ILClone(pidl);
-        return S_OK;
+        hr = S_OK;
+        if (pPidlOut)
+            hr = SHILClone(pidl, pPidlOut);
+        return hr;
```
Comment (HBelusca) [inline dll/win32/shell32/folders/CRegFolder.cpp:778]:
```suggestion
        return (pPidlOut ? SHILClone(pidl, pPidlOut) : S_OK);
```
or, to help with debugging the returned value,
```suggestion
        hr = (pPidlOut ? SHILClone(pidl, pPidlOut) : S_OK);
        return hr;
```

File: dll/win32/shell32/folders/CFSFolder.cpp (hunk 4 above)
Comment (whindsaks) [reply dll/win32/shell32/folders/CFSFolder.cpp:1502]: No, I mean there are pidls with only the ASCII name. Windows still generates them today.

Calling _ILSimpleGetTextW would be less broken but it is still wrong. The correct solution is for this file to interpret its own pidls.

Comment (HBelusca) [reply dll/win32/shell32/folders/CFSFolder.cpp:1502]: Ah, so I understand it's still an unimplemented thing.

File: dll/win32/shell32/folders/CFSFolder.cpp [hunk 6]
```diff
@@ -1499,47 +1499,41 @@ HRESULT WINAPI CFSFolder::SetNameOf(
...
+    else if (MoveFileW(szSrc, szDest))
     {
-        HRESULT hr = S_OK;
-
         if (pPidlOut)
-            hr = _ILCreateFromPathW(szDest, pPidlOut);
-
-        SHChangeNotify (bIsFolder ? SHCNE_RENAMEFOLDER : SHCNE_RENAMEITEM,
-                        SHCNF_PATHW, szSrc, szDest);
+            hr = ParseDisplayName(hwndOwner, NULL, PathFindFileNameW(szDest), NULL, pPidlOut, NULL);
```
Comment (HBelusca) [inline dll/win32/shell32/folders/CFSFolder.cpp:1528]: If the `PathFindFileNameW(szDest)` call returns NULL, is the `ParseDisplayName()` function able to cope with that (i.e. not crash, have a correct behaviour and return a correct error code)?

Comment (whindsaks) [reply dll/win32/shell32/folders/CFSFolder.cpp:1528]: PathFindFileNameW can't return NULL unless the input is NULL and that cannot happen here.

**Pull request merged**